cmd/influxd/run/startup_logger.go

```diff
@@ -0,0 +1,39 @@
+package run
+
+import (
+	"fmt"
+	"sync/atomic"
+
+	"go.uber.org/zap"
+)
+
+type StartupProgressLogger struct {
+	shardsCompleted atomic.Uint64
+	shardsTotal     atomic.Uint64
+	logger          *zap.Logger
+}
+
+func NewStartupProgressLogger(logger *zap.Logger) *StartupProgressLogger {
+	return &StartupProgressLogger{
+		logger: logger,
+	}
+}
+
+func (s *StartupProgressLogger) AddShard() {
+	s.shardsTotal.Add(1)
+}
+
+func (s *StartupProgressLogger) RemoveShardFromCount() {
+	if s.shardsTotal.Load() > 0 {
+		old, newUint := s.shardsTotal.Load(), s.shardsTotal.Load()-1
+		s.shardsTotal.CompareAndSwap(old, newUint)
+	}
```

Comment:
This has a few race conditions in it. All RMW (read-modify-write) operation should only be done with the methods available on [atomic.Uint64](https://pkg.go.dev/sync/atomic#Uint64). Trace two goroutines going through this function at the same time, just one method call on `s.shardsTotal` behind each other and you should see why.

In this case, `s.shardsTotal.Add(-1)` should be used to update it. If you want to check for underflow, you'd check the return value of `Add` and then do `s.shardsTotal.Add(1)` to correct it. Or don't worry about underflow. Things have gone wrong with the usage of `StartupProgressLogger` if you get an underflow.

Reply:
Okay so I should opt for using atomic.Int64 instead and call an `Add(-1)` in this case?